auditbeat: system/process module backed by quark

haesbaert opened the pull request:
## Proposed commit message

This introduces a new provider for the sytem/process module in linux.

The main motivation is to address some of the limitations of the current implementation. The gosysinfo provider sends state reports by scraping /proc from time to time, so it loses all short lived processes. Some customers also would like to have full telemetry but can't run auditd for various reasons.

As a bonus we get some extra ECS fields that were not available before.

MAIN DIFFERENCES:
 * Publishes every process in the system, regardless of lifespan.
 * Publishes exec events for an existing process (without a fork).
 * Aggregates fork+exec+exit within one event.
 * Adds event.exit_code for processes that exited, can't express exit_time in ECS?
 * Include the original process.args, sysinfo reports args that were fetched when it parsed /proc, so a userland process can masquerade itself. For the initial /proc scraping we report the current value like sysinfo. We can't get the original value since the kernel overwrites it, if you wanna have fun: https://github.com/systemd/systemd/blob/main/src/basic/argv-util.c#L165
 * Adds process.args_count.
 * Adds process.interactive and if true, process.tty.char_device.{major,minor}
 * Attempts to hash all processes, not just long lived ones.
 * Hashing is not rate-limited anymore, but it's cached and refreshed based on metadata. It's a LRU keyed by path and refreshed if the metadata of the file changes, statx(2) if the kernel supports, stat(2) otherwise.
 * No more periodic state reports, only initial batch.
 * No more saving the timestamp of the last state-report in disk.
 * No more /proc parsing during runtime, only on boot.

MISSING:
 * Unify entity id with sessionview.
 * Publish metrics from quark.Stats().
 * Docs.
 * Properly define config options and names.

EXTRA CHANGES:
 * Added statx(2) to seccomp_linux so we can properly use CachedHasher.
 * Updated quark to 0.3 so we have namespace inode numbers.

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally

Run auditbeat on linux with the following configuration:
```
auditbeat.modules:

- module: system
  datasets:
    - process
  process.backend: "quark"
```

## Related issues

- https://github.com/elastic/security-integrations/issues/438
- https://github.com/elastic/security-integrations/pull/417

## Screenshots

### Non interactive SSH
Below is a shot of a non interactive ssh session, done with `ssh fc39vm /bin/echo hi from quarkio`.
It shows the intermediary processes of sshd until we fork the shell and echo, the interesting bits is that we can see a process that forked+execed and then execs again: sshd forks+execs mksh,, which in turn execs /bin/echo, without forking.
![ssh_nonint](https://github.com/user-attachments/assets/46fad8dd-f1f9-48cf-bb98-b8c2e56f8a35)

### Comparison against the sysinfo provider for a long lived process:
Here we run a long sleep and just compare the events against the existing provider on 8.14.3:
![vs_old](https://github.com/user-attachments/assets/4cb5fd63-7e3b-4f46-b872-4e4b92fba863)

## On event.type, event.action and others

I've tried to keep things as close as possible to the old provider, but it's really just a suggestion at this point and it's likely we want to change things

TOBECONTINUED

elasticmachine commented:
Pinging @elastic/sec-linux-platform (Team:Security-Linux Platform)

andrewkroh commented on x-pack/auditbeat/module/system/process/config.go:
```diff
@@ -16,11 +17,19 @@ type Config struct {
+	if err := c.HasherConfig.Validate(); err != nil {
+		return err
+	}
+	if c.Backend != "quark" && c.Backend != "proc" {
```
We should be consistent with add_session_metadata in terms of `backend` option names.

haesbaert replied:
Ack, I've changed it to kernel_tracing like that processor, I was hoping we could discuss the naming but it makes more sense to stick with the same for now.

nicholasberlin commented on x-pack/auditbeat/module/system/process/quark_provider_linux.go:
```diff
@@ -0,0 +1,330 @@
+		event.RootFields.Put("process.exit_code", process.Exit.ExitCode)
+	}
+	if !process.Proc.Valid {
+		return event, true
```
`processErr` will be `nil` here, so `makeMessage` in the deferred function will do stuff. Is that what you want? I think the answer is yes, since there is empty username handling. Just checking though.

haesbaert replied:
Correct, makeMessage does handle err being `nil`, it's basically one message if there is an error:
```
func makeMessage(pid int, action eventAction, name string, username string, err error) string {
	if err != nil {
		return fmt.Sprintf("ERROR for PID %d: %v", pid, err)
	}
...
```

nicholasberlin commented on x-pack/auditbeat/module/system/process/quark_provider_linux.go:
```diff
@@ -0,0 +1,330 @@
+	} else {
+		ms.log.Debugf("skipping hash %s (inonum %d vs %d)\n", process.Filename, process.Proc.MntInonum, ms.selfMntNsIno)
+	}
+	event.RootFields.Put("quark", true) // XXX REMOVE ME
```
```suggestion
```
Ready to remove this?

haesbaert replied:
I didn't forget, I swear! :rofl:. I'm keeping this until the last commit, basically it's what I'm using on kibana since I have vanilla auditbeat + auditbeat@quark populating the same table.

haesbaert replied:
oh that's neato, I'll start using that, many thanks.
Removed in ae5a34f17bb1d2c879fdb9b72a00c41e35d04ae3

haesbaert commented on x-pack/auditbeat/module/system/process/quark_provider_linux.go:
```diff
@@ -0,0 +1,333 @@
+	quarkMetrics.backend = monitoring.NewString(reg, "backend", monitoring.Report)
+}
+
+// QuarkMetricSet is a MetricSet with added members used only in by
```
well spotted, thanks!

haesbaert commented on x-pack/auditbeat/module/system/process/quark_provider_linux.go:
```diff
@@ -0,0 +1,333 @@
+	}
+
+	attr := quark.DefaultQueueAttr()
+	qm.queue, err = quark.OpenQueue(attr, 1)
```
It's "how many events can be returned by one future GetEvents call".
I've changed the C code from quark_queue_get_events() to quark_queue_get_event(), to prevent this: https://github.com/elastic/quark/pull/109
That will make 0.4 and will break the API, but the go bindings are still tracking 0.3, so the work around is to just to use `1` which prevents the bad behaviour.
Once 0.4 is out, GetEvents will be zapped in favor of GetEvent, and this argument will be gone.

haesbaert commented on x-pack/auditbeat/module/system/process/quark_provider_linux.go:
```diff
@@ -0,0 +1,333 @@
+	attr := quark.DefaultQueueAttr()
+	qm.queue, err = quark.OpenQueue(attr, 1)
+	if err != nil {
+		return nil, fmt.Errorf("can't open quark queue: %w", err)
```
well spotted, my bad, I moved this around (the queue open was before the cache) and forgot to invert the close.
This is great thanks!

haesbaert replied:
fixed

haesbaert commented on x-pack/auditbeat/module/system/process/quark_provider_linux.go:
```diff
@@ -0,0 +1,333 @@
+
+		x := time.Now()
+		quarkEvents, err := ms.queue.GetEvents()
+		if len(quarkEvents) == 1 {
```
Yes, because we specify `1` in `openQueue`, GetEvents cannot be > 1.
Initially I was using this to gauge (which I can pronounce now) how long GetEvents took, and it was easier to reason by looking just at the cases where it's a single event, otherwise I'd have to divide and whatnot.

haesbaert commented:
This is ready for some bashing, major changes since the initial push:

- Added metrics, naming and what should be gauge needs consensus.
- Added tests for ebpf and kprobe, for initial snapshot and generated events.
- process.backend is now `procfs` or `kernel_tracing`, instead of `proc` and `quark`.
- Still missing consensus on event.action and event.type.

nicholasberlin commented:
> Still missing consensus on event.action and event.type.

I think we should treat event.action as an array, as Endpoint is currently doing.

nicholasberlin reviewed (commented):
LGTM so far.

haesbaert commented:
> > Still missing consensus on event.action and event.type.
> 
> I think we should treat event.action as an array, as Endpoint is currently doing.

That'd be great, it would simplify a lot.

nfritts commented:
@andrewkroh any further thoughts on this? 

@haesbaert this will be a new optional backend that isn't used by default after this PR right?   If that's the case it feels fairly safe to me if we merge it and look in to have some additional parties test it.

andrewkroh commented on x-pack/auditbeat/module/system/process/process.go:
```diff
@@ -84,12 +94,21 @@ func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+		if qm, err := NewFromQuark(ms); err == nil {
+			return qm, nil
+		} else {
+			ms.log.Errorf("can't use kernel_tracing, falling back to procfs: %w", err)
```
You cannot use `%w` with `logp.Errorf`. That's only for wrapping errors with `fmt.Errorf`.

Switch to `%v` or change to a structured log message like `ms.log.Errorw("my static message", "error", err)` (which will render as a JSON object containing the `err.Error()` value in the ECS `error.message` field).

haesbaert replied:
well spotted, many thanks!

andrewkroh commented on x-pack/auditbeat/module/system/process/process.go:
```diff
@@ -102,3 +121,31 @@ func entityID(hostID string, pid int, startTime time.Time) string {
+	}
+
+	var actionString string
+	switch action {
```
This switch is missing `eventActionExistingProcess`. Is that intentional?

haesbaert replied:
it's there on line 136, did you mean `eventActionProcessError`?

andrewkroh replied:
Sorry, I must have copy/pasted the wrong enum item. Yeah, I meant `eventActionProcessError`.

haesbaert replied:
Well spot, there is a difference indeed:
It's actually non-reachable code, the gosysinfo provider only looks at `process.Error` and thus never sees `eventActionProcessError`, but it does set action to `eventActionProcessError`.

quark never actually sets `eventActionProcessError`, but sets `processErr`.

I was trying to make them both match, but I believe the behavior of gosysinfo here is misleading, or at least on the hashing case:
When the hashing fails, gosysinfo will set action to `eventActionProcessError`, but I don't think it should, I don't think we should clobber `action` just because the file is too big and we can't hash, we could still include `error.message`.

Having said that, I'll change quark to have the same behavior as gosysinfo, and then later we can think about changing both.

haesbaert replied:
I'm also not a big fan of all the upper case on error messages, it looks like we're screaming at the user, but not something for this PR.

haesbaert replied:
Handled it even, if not reached (for now): d04023488104043f56b622e23b4f5b3eb19d1cf6

andrewkroh commented on x-pack/auditbeat/module/system/process/quark_provider_linux_test.go:
```diff
@@ -0,0 +1,411 @@
+}
+
+// makeSelfEvent builds what should be the event that quark will
+// generate as a initial snapshot of the current process
```
```suggestion
// generate as an initial snapshot of the current process
```

andrewkroh reviewed (commented):
Nice work. Excited to have this capability available.

Just a few minor comments.

haesbaert commented on x-pack/auditbeat/module/system/process/quark_provider_linux.go:
```diff
@@ -0,0 +1,338 @@
+		if err != nil {
+			processErr = fmt.Errorf("failed to hash executable %v for PID %v: %w",
+				process.Filename, process.Pid, err)
+			ms.log.Error(processErr.Error())
```
Indeed, I was just keeping parity with the gosysinfo code that `ms.log.Warn` here: https://github.com/elastic/beats/blob/4681030c67054bcbe0715a11f4056d245b1dd84f/x-pack/auditbeat/module/system/process/gosysinfo_provider.go#L182

It should have been a Warn, I don't feel strongly about it, especially since this is a somewhat common warning/error, as the default size limit is not huge.

andrewkroh replied:
Then let's lower to `Warn`.

haesbaert replied:
fixed in 2bb1815a7979143a0d4b78900315c880ec441b32

elasticmachine commented:
Pinging @elastic/elastic-agent-data-plane (Team:Elastic-Agent-Data-Plane)

andrewkroh commented on filebeat/channel/runner_mock_test.go:
```diff
@@ -69,17 +70,19 @@ func (r runnerFactoryMock) Assert(t *testing.T) {
-				require.NotSamef(t, c1.Processing.Meta, c2.Processing.Meta, "`Processing.Meta` cannot be re-used")
-				require.NotSamef(t, c1.Processing.Fields, c2.Processing.Fields, "`Processing.Fields` cannot be re-used")
-				require.NotSamef(t, c1.Processing.Processor, c2.Processing.Processor, "`Processing.Processor` cannot be re-used")
+				require.Falsef(t, sameBacking(c1.Processing.Meta, c2.Processing.Meta), "`Processing.Meta` cannot be re-used")
```
I think we accomplish the same thing (without needing unsafe pointer) by passing map pointers into `NotSame` . e.g.

```
				require.NotSamef(t, &c1.Processing.Meta, &c2.Processing.Meta, "`Processing.Meta` cannot be re-used")
				require.NotSamef(t, &c1.Processing.Fields, &c2.Processing.Fields, "`Processing.Fields` cannot be re-used")
```

haesbaert replied:
I might be missing something, but I think it does not. I think the addresses of the maps are not the addresses of the backing storage of said map, I'm not super versed in go so I might be misinterpreting it:
```
	x := make(map[string]int)
	x["foo"] = 2
	y := x
	y["bar"] = 3

	require.NotSame(t, &x, &y, "lalalal") <--- should blow up here, but doesn't
	fmt.Printf("same backing = %v\n", sameBacking(x, y))
	require.Falsef(t, sameBacking(x, y), "`Processing.Meta` cannot be re-used") <--- blows up here
```
Here sameBacking(x, y) is true, but require.NotSame(&x, &y) is also true (so it passes when it shouldn't)

andrewkroh replied:
You are right. This needs the underlying map data to be correct.

haesbaert commented:
I'm more or less ready to commit this, as there were some main merges since I did the heavy testing I just want to redo them.
Basically I let it run for a day on fork bombs to check if it behaves.

AndersonQ reviewed (commented):
@haesbaert is there any change you forgot to commit a file? I tried to run it, but it does not accept the `quark` backend:

```
{"log.level":"error","@timestamp":"2025-02-20T16:05:54.355+0100","log.origin":{"function":"github.com/elastic/beats/v7/libbeat/cmd/instance.handleError","file.name":"instance/beat.go","file.line":1589},"message":"Exiting: 1 error: failed to unpack the system/process config: invalid process.backend 'quark' accessing 'auditbeat.modules.2' (source:'auditbeat-anderson.yml')","service.name":"auditbeat","ecs.version":"1.6.0"}
```

I had a quick look and it seems the config does not accept the quark backed

https://github.com/elastic/beats/blob/d3285d467c6cea672d4216d72cb36f39c1eb71ad/x-pack/auditbeat/module/system/process/config.go#L23-L33


this is the config I used:

```yaml
http:
  enabled: true

auditbeat.modules:

- module: system
  datasets:
    - process
  process.backend: "quark"

path.home: /tmp/beat

output.file:
  path: /tmp/beat/
  filename: auditbeat-output-file
  rotate_every_kb: 10000

processors:
  - add_host_metadata: ~
  - add_cloud_metadata: ~
  - add_docker_metadata: ~

logging.level: info
monitoring.enabled: true
logging.metrics:
  - enabled:

```

AndersonQ reviewed (changes requested):
just until we clarify it runs. as you can see in my comment above, I cannot run it

haesbaert commented on x-pack/auditbeat/module/system/process/quark_provider_linux_test.go:
Ack, that was intentional, most of the tests only make sense if the rest didn't fail, for example, no point in checking:
```
	require.True(t, qev.Process.Proc.Valid)
if the test above failed:
	require.Equal(t,
		qev.Events&(quark.QUARK_EV_FORK|quark.QUARK_EV_EXEC|quark.QUARK_EV_EXIT),
		quark.QUARK_EV_FORK|quark.QUARK_EV_EXEC|quark.QUARK_EV_EXIT)
```

haesbaert commented:
> just until we clarify it runs. as you can see in my comment above, I cannot run it

the keyword changed to "kernel_tracing" instead of "quark" for the backend, so if you change your config it should work.

https://github.com/elastic/beats/pull/42032/commits/e3b733227a637964db13018175381e7fa6053934
https://github.com/elastic/beats/pull/42032#discussion_r1884355376

haesbaert commented:
> @haesbaert is there any change you forgot to commit a file? I tried to run it, but it does not accept the `quark` backend:
> 
> ```
> {"log.level":"error","@timestamp":"2025-02-20T16:05:54.355+0100","log.origin":{"function":"github.com/elastic/beats/v7/libbeat/cmd/instance.handleError","file.name":"instance/beat.go","file.line":1589},"message":"Exiting: 1 error: failed to unpack the system/process config: invalid process.backend 'quark' accessing 'auditbeat.modules.2' (source:'auditbeat-anderson.yml')","service.name":"auditbeat","ecs.version":"1.6.0"}
> ```
> 
> I had a quick look and it seems the config does not accept the quark backed
> 
> https://github.com/elastic/beats/blob/d3285d467c6cea672d4216d72cb36f39c1eb71ad/x-pack/auditbeat/module/system/process/config.go#L23-L33
> 
> this is the config I used:
> 
> ```yaml
> http:
>   enabled: true
> 
> auditbeat.modules:
> 
> - module: system
>   datasets:
>     - process
>   process.backend: "quark"
> 
> path.home: /tmp/beat
> 
> output.file:
>   path: /tmp/beat/
>   filename: auditbeat-output-file
>   rotate_every_kb: 10000
> 
> processors:
>   - add_host_metadata: ~
>   - add_cloud_metadata: ~
>   - add_docker_metadata: ~
> 
> logging.level: info
> monitoring.enabled: true
> logging.metrics:
>   - enabled:
> ```

Correct, it changed to "kernel_tracing", it will try quark on ebpf, if it fails, it will try quark on kprobe, if it fails it fallsback to gosysinfo.

haesbaert commented on x-pack/auditbeat/module/system/process/process.go:
```diff
@@ -102,3 +121,33 @@ func entityID(hostID string, pid int, startTime time.Time) string {
+		userString = fmt.Sprintf(" by user %v", username)
+	}
+
+	return fmt.Sprintf("Process %v (PID: %d)%v %v",
+		name, pid, userString, actionString)
```
Not really my code, this is what the processor was already doing, I just unified the call:
https://github.com/elastic/beats/blob/main/x-pack/auditbeat/module/system/process/gosysinfo_provider.go#L373-L374

AndersonQ commented:
> ## How to test this PR locally

ah, ok. could you please update the "How to test this PR" section?

haesbaert commented:
> > ## How to test this PR locally
> 
> ah, ok. could you please update the "How to test this PR" section?

zefixed, thanks for testing :)

haesbaert commented:
This is the output of valgrind after ~6h of quark-mon, running together with auditbeat, just to make sure there's nothing wrong in the C side and a reminder to self of how things were.
```
==47391== HEAP SUMMARY:
==47391==     in use at exit: 0 bytes in 0 blocks
==47391==   total heap usage: 5,326,567 allocs, 5,326,567 frees, 4,295,175,100 bytes allocated
```

AndersonQ reviewed (approved):
LGTM, I tested and ti works :)
I left some comments, but I don't see them as blockers. It's up to you to decide if they're relevant or not

haesbaert commented on x-pack/auditbeat/module/system/process/quark_provider_linux.go:
```diff
@@ -0,0 +1,337 @@
+		if err != nil {
+			processErr = fmt.Errorf("failed to hash executable %v for PID %v: %w",
+				process.Filename, process.Pid, err)
+			ms.log.Warn(processErr.Error())
```
I agree!
I want to revamp the error messages, by now I'm just keeping really 1:1 with the former backend.
It's pretty annoying to get that, it would be nice if we can warn it somehow in another place, or just supress it.
